src/Adapter/MSTest.TestAdapter/Execution/TestMethodRunner.cs

```diff
@@ -254,6 +249,46 @@ internal UnitTestResult[] RunTestMethod()
         return results.ToUnitTestResults();
     }
 
+    private bool TryExecuteITestDataSource(List<TestResult> results)
+    {
+        if (_test.DataType != DynamicDataType.ITestDataSource)
+        {
+            return false;
+        }
+
+        UTF.ITestDataSource? dataSource;
+        object?[]? data;
+        if (_test.SerializedData?.Length == 3
+            && Enum.TryParse(_test.SerializedData[0], out TestDataIdentifierStrategy _))
+        {
+            if (!int.TryParse(_test.SerializedData[1], out int dataSourceIndex)
+                || !int.TryParse(_test.SerializedData[2], out int dataIndex))
+            {
+                // TODO
+                throw new InvalidOperationException();
```

Comment:
Shall I handle gracefully or throw unreachable?

Reply:
I think this should be unreachable. Otherwise something very bad happened.

If SerializedData is coming through data contract serialization, the length must be even. So, we are sure if the length is 3, it should be coming through the new alternative. In this case, we should always be able to correctly parse everything back.

I'd even move the `Enum.TryParse` check from the outer if to the inner if (negated, of course)